core/src/main/scala/org/apache/spark/BarrierCoordinator.scala

```diff
@@ -51,7 +52,8 @@ private[spark] class BarrierCoordinator(
 
   // TODO SPARK-25030 Create a Timer() in the mainClass submitted to SparkSubmit makes it unable to
   // fetch result, we shall fix the issue.
-  private lazy val timer = new Timer("BarrierCoordinator barrier epoch increment timer")
+  private lazy val timer = ThreadUtils.newSingleThreadScheduledExecutor(
```

Comment:
Doesn't cancelTimerTask() have to change too?

Reply:
The `purge` of `Timer` has the same name with the purge of `ThreadScheduledExecutor`.

Reply:
OK. I wonder if onStop() should also shutdown the timer? it isn't in the current code, but maybe it would be better to ensure it is stopped there

Reply:
Good idea.